src/dynamics/mpas/driver/dyn_mpas_subdriver.F90

```diff
@@ -1264,70 +1590,141 @@ subroutine add_stream_attribute_1d(attribute_name, attribute_value)
 
-        character(64), allocatable :: var_name_list(:)
-        integer :: i, n, offset
-        type(var_info_type), allocatable :: var_info_list_append(:)
+        character(*), parameter :: supported_stream_name_operator = '+-'
+        character(1) :: stream_name_operator
+        character(:), allocatable :: stream_name_fragment
+        character(len(invariant_var_info_list % name)), allocatable :: var_name_list(:)
+        integer :: i, j, n, offset
+        type(var_info_type), allocatable :: var_info_list_buffer(:)
 
-        allocate(var_info_list(0))
+        n = len_trim(stream_name)
 
-        n = len(stream_name)
-        offset = 0
+        if (n == 0) then
+            ! Empty character string means empty list.
+            var_info_list = parse_stream_name_fragment('')
+
+            return
+        end if
+
+        i = scan(stream_name, supported_stream_name_operator)
+
+        if (i == 0) then
+            ! No operators are present in the stream name. It is just a single stream name fragment.
+            stream_name_fragment = stream_name
+            var_info_list = parse_stream_name_fragment(stream_name_fragment)
 
-        if (offset + 1 > n) then
             return
         end if
 
-        i = index(stream_name(offset + 1:), '+')
+        offset = 0
+        var_info_list = parse_stream_name_fragment('')
+
+        do while (.true.)
+            ! Extract operator from the stream name.
+            if (offset > 0) then
+                stream_name_operator = stream_name(offset:offset)
+            else
+                stream_name_operator = '+'
+            end if
 
-        do while (i > 0)
+            ! Extract stream name fragment from the stream name.
             if (i > 1) then
-                var_info_list_append = parse_stream_name(stream_name(offset + 1:offset + i - 1))
-                var_info_list = [var_info_list, var_info_list_append]
+                stream_name_fragment = stream_name(offset + 1:offset + i - 1)
+            else
+                stream_name_fragment = ''
+            end if
 
-                deallocate(var_info_list_append)
+            ! Process the stream name fragment according to the operator.
+            if (len_trim(stream_name_fragment) > 0) then
+                var_info_list_buffer = parse_stream_name_fragment(stream_name_fragment)
+
+                select case (stream_name_operator)
+                    case ('+')
+                        var_info_list = [var_info_list, var_info_list_buffer]
+                    case ('-')
+                        do j = 1, size(var_info_list_buffer)
+                            var_name_list = var_info_list % name
+                            var_info_list = pack(var_info_list, var_name_list /= var_info_list_buffer(j) % name)
+                        end do
+                    case default
+                        ! Do nothing for unknown operators. Should not happen at all.
```

Comment:
The quoted code section here is inside a **pure** function, `parse_stream_name`. We cannot have I/O statements in pure procedures.

The debuggability should still be fine though. This function returns a list (rank-1 array) of `var_info_type`. Wherever it is called, there are debug messages below it like [here](https://github.com/ESCOMP/CAM-SIMA/blob/724d5c859556d4ed9bc13de039b574f0810b7252/src/dynamics/mpas/driver/dyn_mpas_subdriver.F90#L1294-L1305), which prints the elements of that list. A bad operator will cause the resulting list to either contain extra or missing elements. Therefore, by inspecting debug messages, this should be fairly obvious to spot.

Reply:
Ahh, yeah, I missed that it was in a pure function.  I agree that doing the debug-related checks after the function has returned is fine.  

If we ever find that this is a common problem (which I doubt it will be) then we can always have it return very specific quantities in `var_info_type` that is then checked immediately afterwards, but I certainly wouldn't worry about it for this PR (if ever).